libs/platform/ebpf_hash_table.c

```diff
@@ -111,125 +242,227 @@ ebpf_result_t
+    // if the key is already present.
+    // Note: This assumes either the table is synchronized by an external lock
+    // or that memory is freed when the epoch ends.
+    for (;;) {
+        // Capture the value of the list.
+        previous_bucket = &(hash_table->buckets[hash % hash_table->bucket_count]);
+        first_bucket = hash_table->buckets[hash % hash_table->bucket_count];
+
+        // Check the captured list to see if it already contains this key.
+        old_bucket = _ebpf_hash_table_find_bucket(hash_table, &previous_bucket, key);
+
+        // Atomically replace old bucket with new bucket.
+        if (old_bucket) {
+            // Only replace if it hasn't changed.
+            if (ebpf_interlocked_compare_exchange_pointer((void* volatile*)previous_bucket, new_bucket, old_bucket) !=
+                old_bucket)
+                continue;
+
+        } else { // Atomically insert at head of list.
+            // Only replace if it hasn't changed.
+            new_bucket->next = first_bucket;
+            if (ebpf_interlocked_compare_exchange_pointer(
+                    (void* volatile*)&(hash_table->buckets[hash % hash_table->bucket_count]),
+                    new_bucket,
+                    first_bucket) != first_bucket) {
+                continue;
+            }
+        }
+        // Insert was successful
+        break;
     }
 
-    // Update existing entry
-    if (value)
-        memcpy(entry + hash_table->key_size, value, hash_table->value_size);
+    // Free the old bucket.
+    hash_table->free(old_bucket);
 
-    retval = EBPF_SUCCESS;
+    ebpf_interlocked_increment_int32((volatile int32_t*)&hash_table->entry_count);
 
 Done:
-    ebpf_free(temp);
-
     return retval;
 }
 
 ebpf_result_t
 ebpf_hash_table_delete(_In_ ebpf_hash_table_t* hash_table, _In_ const uint8_t* key)
 {
-    BOOLEAN result;
-    RTL_AVL_TABLE* table = (RTL_AVL_TABLE*)hash_table;
+    ebpf_result_t retval;
+    uint32_t hash;
+    ebpf_hash_bucket_header_t* bucket;
+    ebpf_hash_bucket_header_t** previous_bucket;
+
+    if (!hash_table || !key) {
+        retval = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    hash = _ebpf_hash_table_compute_hash(hash_table, key);
+
+    // Atomically remove the bucket.
+    for (;;) {
+        // Capture the start of the bucket chain.
+        previous_bucket = hash_table->buckets + hash % hash_table->bucket_count;
+
+        // Find the location of the bucket containing this key.
+        bucket = _ebpf_hash_table_find_bucket(hash_table, &previous_bucket, key);
+
+        if (bucket == NULL) {
+            retval = EBPF_KEY_NOT_FOUND;
+            goto Done;
+        }
+
+        // Set the previous bucket's next value to bucket next value only if
+        // previous bucket's next value has not changed.
+        if (ebpf_interlocked_compare_exchange_pointer((void* volatile*)previous_bucket, bucket->next, bucket) !=
```

Comment:
Alan will check for correctness here. Likely a bug if a->b->c->d (b,c) are getting deleted concurrently.